hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/cache/EntityMapWarmer.java

```diff
@@ -0,0 +1,258 @@
+    private final ThreadPoolExecutor threadpool;
+
+    private EntityMapWarmer(
+            Supplier<AccountStorageAdapter> accountsStorageAdapter,
+            Supplier<UniqueTokenMapAdapter> nftsAdapter,
+            Supplier<TokenRelStorageAdapter> tokenRelsAdapter,
+            GlobalDynamicProperties globalDynamicProperties) {
+        this(
+                accountsStorageAdapter,
+                nftsAdapter,
+                tokenRelsAdapter,
+                new ThreadPoolExecutor(
+                        globalDynamicProperties.cacheCryptoTransferWarmThreads(),
+                        globalDynamicProperties.cacheCryptoTransferWarmThreads(),
+                        0L,
+                        TimeUnit.MILLISECONDS,
+                        new LinkedBlockingQueue<>()));
+    }
+
+    @VisibleForTesting
+    EntityMapWarmer(
+            Supplier<AccountStorageAdapter> accountsStorageAdapter,
+            Supplier<UniqueTokenMapAdapter> nftsAdapter,
+            Supplier<TokenRelStorageAdapter> tokenRelsAdapter,
+            ThreadPoolExecutor threadpool) {
+        this.accountsStorageAdapter = accountsStorageAdapter;
+        this.nftsAdapter = nftsAdapter;
+        this.tokenRelsAdapter = tokenRelsAdapter;
+        this.threadpool = threadpool;
+    }
+
+    public static EntityMapWarmer getInstance(
+            Supplier<AccountStorageAdapter> accountsStorageAdapter,
+            Supplier<UniqueTokenMapAdapter> nftsAdapter,
+            Supplier<TokenRelStorageAdapter> tokenRelsAdapter,
+            GlobalDynamicProperties globalDynamicProperties) {
+        if (instance == null) {
+            instance =
+                    new EntityMapWarmer(accountsStorageAdapter, nftsAdapter, tokenRelsAdapter, globalDynamicProperties);
+        }
+        return instance;
+    }
+
+    /**
+     * Warms the cache for select entities found in the given {@code Round}. Any entities not
+     * found will, of course, not be warmed, but instead will essentially be ignored.
+     *
+     * @param round the round to warm the cache for
+     */
+    public void warmCache(Round round) {
+        removeAllTasks();
+        threadpool.execute(() -> doWarmups(round));
+    }
+
+    // "warmTokenObjs" as in "warm the token and its associated token relation objects"
+    // for both the sending and receiving accounts
+    private void warmTokenObjs(AccountID sender, AccountID receiver, long tokenNum, long senderSerialNum) {
+        if (accountsStorageAdapter.get().areOnDisk()
+                && nftsAdapter.get().isVirtual()
+                && tokenRelsAdapter.get().areOnDisk()) {
+
+            threadpool.execute(() -> {
+                // Sender:
+                final var acctMap = accountsStorageAdapter.get().getOnDiskAccounts();
+                final OnDiskAccount account =
+                        Objects.requireNonNull(acctMap).get(EntityNumVirtualKey.from(EntityNum.fromAccountId(sender)));
+                if (account != null) {
```

Comment:
I understand why we'd warm the "root node" of the **`receiver`**'s  `ownedNfts` linked list, since we'll insert its new acquisition at the root.

But unless the `sender` is transferring its most recently acquired NFT (i.e., unless `headSerialNum == senderSerialNum`), this is wasted work.

Reply:
Hmm, it might be that I over-interpreted observed behavior when a 100M-NFT benchmark transfers its first 100M NFTs, which takes hours, from their treasury accounts (would that match your condition?).
I consider this warming code WIP and will revisit/reevaluate it with your suggestion in mind.

Reply:
@tinker-michaelj @OlegMazurov what would you like to do about the code in its current state? Discuss? Merge as is? Change and test? Something else? Not sure what the designation of _WIP_  means for this PR

Reply:
Merge as is. I was able to demonstrate the benefit of the current implementation for a particular benchmark. I now realize some parts of the warming code may be considered _ad hoc_. With more changes to come to help improve NFT transfer performance, I'll  continue evaluating the effect and overhead of this code (i.e. it's a "work in progress" for me, not for this PR).